File: Data Structures & Algorithms/squares-of-a-sorted-array/submission-1.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        // The input array is
        // [-4, -1, 0, 3, 10]
        vector<int> res;
        int left = 0, right = nums.size() - 1;

        while (left <= right) {
            if (nums[left] * nums[left] > nums[right] * nums[right]) {
                res.push_back(nums[left] * nums[left]);
                left++;
            } else {
                res.push_back(nums[right] * nums[right]);
                right--;
            }
        }
        // Right now the res looks like:
        // [100, 16, 9, 1, 0]
        // We have to reverse the res array so the array is sorted in non-decreasing order
        reverse(res.begin(), res.end()); 
        return res;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        n = len(nums)
        for i, num in enumerate(nums):
            if num > 0: break
            if (i > 0 and nums[i] == nums[i - 1]): continue
            left = i + 1
            right = n - 1
            while left < right:
                threeSum = num + nums[left] + nums[right]
                if threeSum > 0: right -= 1
                elif threeSum < 0: left += 1
                else:
                    res.append([num, nums[left], nums[right]])
                    left += 1
                    right -= 1
                    while (left < right and nums[left] == nums[left - 1]):
                        left += 1
        return res